IdentityCore/src/requests/MSIDSilentTokenRequest.m

```diff
@@ -239,41 +237,46 @@ - (void)fetchCachedRefreshTokenWithCompletionHandler:(void (^)(MSIDBaseToken<MSI
     }
     
     NSError *rtError = nil;
-    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"Looking for app refresh token...");
-    MSIDBaseToken<MSIDRefreshableToken> *refreshableToken = [self appRefreshTokenWithError:&rtError];
+    MSIDBaseToken<MSIDRefreshableToken> *refreshableToken = nil;
+    NSString *contextMsg = checkForFRT ? @"family refresh token" : @"app refresh token";
+    MSIDRefreshTokenTypes checkForTokenType = checkForFRT ? MSIDFamilyRefreshTokenType : MSIDAppRefreshTokenType;
     
-    if (rtError)
+    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"Looking for %@...", contextMsg);
+    if (checkForFRT)
     {
-        MSID_LOG_WITH_CTX_PII(MSIDLogLevelError, self.requestParameters, @"Failed to read app specific refresh token with error %@", MSID_PII_LOG_MASKABLE(rtError));
-        completionHandler(nil, MSIDAppRefreshTokenType, rtError);
-        return;
-    }
-    
-    if (!refreshableToken)
-    {
-        MSID_LOG_WITH_CTX_PII(MSIDLogLevelWarning, self.requestParameters, @"Didn't find app refresh token");
+        refreshableToken = [self familyRefreshTokenWithError:&rtError];
     }
     else
     {
-        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"Found app refresh token.");
-        completionHandler(refreshableToken, MSIDAppRefreshTokenType, nil);
-        return;
+        refreshableToken = [self appRefreshTokenWithError:&rtError];
     }
     
-    // If no ART, get family refresh token instead.
-    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"Looking for family refresh token...");
-    refreshableToken = [self familyRefreshTokenWithError:&rtError];
-    
     if (rtError)
     {
-        MSID_LOG_WITH_CTX_PII(MSIDLogLevelError, self.requestParameters, @"Failed to read family refresh token with error %@", MSID_PII_LOG_MASKABLE(rtError));
-        completionHandler(nil, MSIDFamilyRefreshTokenType, rtError);
+        MSID_LOG_WITH_CTX_PII(MSIDLogLevelError, self.requestParameters, @"Failed to read %@ token with error %@", contextMsg, MSID_PII_LOG_MASKABLE(rtError));
+        completionHandler(nil, checkForTokenType, rtError);
         return;
     }
     
-    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"Found family refresh token.");
-    completionHandler(refreshableToken, MSIDFamilyRefreshTokenType, nil);
-    return;
+    if (!refreshableToken)
+    {
+        // Handle to continue or complete
+        MSID_LOG_WITH_CTX_PII(MSIDLogLevelWarning, self.requestParameters, @"Didn't find %@", contextMsg);
+        if (!shouldComplete)
+        {
+            [self fetchCachedTokenAndCheckForFRTFirst:!checkForFRT shouldComplete:!shouldComplete completionHandler:completionHandler];
+        }
+        else
+        {
+            // If no refreshableToken was found, simply return nil as token and token type that the last one it tries to find
```

Comment:
Can you confirm my understanding that this is correct:
If no FRT was found for silent cache lookup, this will return nil even if there was an MRRT in cache. Is this expected?

Reply:
Thanks for pointing out this one. I read the original logic (before the ART and FRT reverting)

1. If the AT is still valid, we will try to find FRT first than MRRT (error doesn't matter)
2. If we need to do a silent call, we will try to find FRT first from cache, and if there is an error, will complete with nil without checking the MRRT 

Should we keep the #1, #2 or keep them separately?

Reply:
Discussed offline - my concern was that we don't return RT at all if FRT not found (I've misread the code). 
But if there was a keychain error, we might want to continue lookup regardless.

Reply:
Agree with this I will update the logic a bit to support continuing lookup if there is keychain error

Reply:
Could you clarify ```keychain error``` mean? Is it a specific error type I need to verify or a general error (rtError in the code) during cache lookup?

Reply:
I think general error is fine, because all errors in this method should be keychain errors. (Keychain error is some kind of an error happening during keychain read operations)

Reply:
I have updated the logic and if should not complete and there is an error, will continue lookup